cli/cmd/docker/builder.go

```
@@ -169,3 +177,169 @@ func AddDockerfileToBuildContext(dockerfileCtx io.ReadCloser, buildCtx io.ReadCl
+		"--tag", fmt.Sprintf("%s:%s", opts.ImageRepo, opts.Tag),
+		"--cache-from", fmt.Sprintf("%s:%s", opts.ImageRepo, opts.CurrentTag),
+	}
+	for key, val := range opts.Env {
+		commandArgs = append(commandArgs, "--build-arg", fmt.Sprintf("%s=%s", key, val))
+	}
+
+	commandArgs = append(commandArgs, extraDockerArgs...)
+	// note: the path _must_ be the last argument
+	commandArgs = append(commandArgs, opts.BuildContext)
+
+	stdoutWriters := []io.Writer{os.Stdout}
+	stderrWriters := []io.Writer{os.Stderr}
+	if opts.LogFile != nil {
+		stdoutWriters = append(stdoutWriters, opts.LogFile)
+		stderrWriters = append(stderrWriters, opts.LogFile)
+	}
+
+	// #nosec G204 - The command is meant to be variable
+	cmd := exec.CommandContext(ctx, "docker", commandArgs...)
+	cmd.Dir = opts.BuildContext
+	cmd.Stdout = io.MultiWriter(stdoutWriters...)
+	cmd.Stderr = io.MultiWriter(stderrWriters...)
+	if err := cmd.Start(); err != nil {
+		return fmt.Errorf("unable to start the build command: %w", err)
+	}
+
+	exitCode := 0
+	execErr := cmd.Wait()
+	if execErr != nil {
+		if exitError, ok := execErr.(*exec.ExitError); ok {
+			exitCode = exitError.ExitCode()
+		}
+	}
+
+	if ctx.Err() == context.Canceled {
+		return fmt.Errorf("build command canceled: %w", ctx.Err())
+	}
+
+	if err := ctx.Err(); err != nil {
+		return fmt.Errorf("error while running build: %w", err)
+	}
+
+	if exitCode != 0 {
+		return fmt.Errorf("build exited with non-zero exit code %d", exitCode)
+	}
+
+	return nil
+}
+
+func injectDockerfileIntoBuildContext(buildContext string, dockerfilePath string) error {
+	randomName := ".dockerfile." + stringid.GenerateRandomID()[:20]
+	data := map[string]func() ([]byte, error){
+		randomName: func() ([]byte, error) {
+			dockerfilePath = filepath.Clean(dockerfilePath)
+			dockerfileCtx, err := os.Open(dockerfilePath)
+			if err != nil {
+				return []byte{}, err
+			}
+			defer func() {
+				if err := dockerfileCtx.Close(); err != nil {
+					log.Fatalln(err)
```

Does this need to be fatal? Can we bubble up an error to handle at the top level?